### PR TITLE
[PEPPER-1370] Do not export NOT_SPECIFIED medical records

### PR DESCRIPTION
**Pull request by denniscunningham**
## Context
See PEPPER-1370 for an explanation of the situation that this PR addresses, but the meta issue is the discrepancy between the medical records DSM writes to ES when they are created versus the medical records it exports to ES.

Note that I actually implemented each of the solutions mentioned in PEPPER-1370, mostly because each approach had unexpected side effects. This PR therefore includes remnants of those changes, largely in `MedicalRecord`. Since those changes made the code clearer and more flexible, I left them.

The primary change is to stop writing medical records that are only present to support `OncHistoryDetail` records. (This is an unfortunate side effect of how institutions are modeled.) Those changes are in `MedicalRecordUtil`, and the testing support is in `ElasticExportServiceTest`, with updates to `OsteoMigratorTest` to adjust for the change. These medical records are associated with an institution of type `NOT_SPECIFIED`, so the code uses that term as a reference.

The secondary change is to modify how DSM identifies ES documents when changing the destruction policy in all `OncHistoryDetail` records for a realm. Those changes are in `OncHistoryDetail. updateDestructionPolicy`, and the testing support is in `OncHistoryDetailPatchTest`. DSM was relying on `NOT_SPECIFIED` medical records being present in ES for the implementation, so it now gets that information from `Participant` records.

Other changes in this PR include:

1.  Took another pass at fixing the code in `OncHistoryDetailPatch`, since I improved `OncHistoryDetail.verifyOrCreateMedicalRecord` and we now have test coverage.
2. Stopped passing a string version of the integer instance ID in `MedicalRecordUtil`, `DDPMedicalRecordDataRequest`, etc.
3. Added an osteo ES mapping file for testing. I was using inappropriate mappings for some test indexes, and ran into a test error due to that.
4. Ensure `Participant` objects include an instance ID.
5. Refactored the SQL queries in `MedicalRecord` to be clearer about what predicates apply to each of the 'get' methods. 
6. Improved a few methods in `MedicalRecordUtil` to support simplifying `OncHistoryDetail.verifyOrCreateMedicalRecord`.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have added regression test labels to the jira ticket
- [ ] I have added verification steps to the jira ticket
- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/MedicalRecord.java [hunk 1]
```diff
@@ -296,47 +313,37 @@ public static Map<String, List<MedicalRecord>> getMedicalRecords(@NonNull String
...
             } catch (SQLException e) {
                 throw new DsmInternalError("Error getting list of medical records for participant " + participantId, e);
             }
-            logger.info("Got {} medical records for participant {}", medicalRecords.size(), participantId);
+            log.info("Got {} medical records for participant {}", medicalRecords.size(), participantId);
             return medicalRecords;
         });
     }
 
-    public static Map<String, List<MedicalRecord>> getMedicalRecordsByParticipantIds(@NonNull String realm, List<String> participantIds) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/MedicalRecord.java:335]: This method was unused.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistoryDetail.java [hunk 2]
```diff
@@ -230,46 +228,11 @@ public class OncHistoryDetail implements HasDdpInstanceId {
     public OncHistoryDetail() {
     }
 
+    @VisibleForTesting
     public OncHistoryDetail(long ddpInstanceId) {
         this.ddpInstanceId = ddpInstanceId;
     }
 
-    public OncHistoryDetail(Integer oncHistoryDetailId, Integer medicalRecordId, String datePx, String typePx, String locationPx,
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistoryDetail.java:237]: Unused.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistoryDetail.java [hunk 3]
```diff
@@ -350,7 +313,7 @@ public static OncHistoryDetail getOncHistoryDetail(@NonNull ResultSet rs) throws
     }
 
     // TODO: there should be no need for the realm parameter -DC
-    public static OncHistoryDetail getOncHistoryDetail(@NonNull int oncHistoryDetailId, String realm) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/OncHistoryDetail.java:353]: `@NonNull` makes no sense on atomics, and in general add no/little value. If a parameter value can be null, that should be part of the method JavaDocs.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/Participant.java [hunk 4]
```diff
@@ -144,29 +145,10 @@ public Participant(Long participantId, String ddpParticipantId, String assigneeI
         this.exitDate = exitDate;
     }
 
-    public Participant(Long participantId, String ddpParticipantId, String assigneeIdMr, Integer ddpInstanceId, String assigneeIdTissue,
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/Participant.java:147]: Unused.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/Participant.java [hunk 5]
```diff
@@ -190,48 +172,42 @@ public static Participant getParticipant(@NonNull Map<String, Assignee> assignee
...
+            rs.getString(DBConstants.DDP_PARTICIPANT_RECORD_ALIAS + DBConstants.ALIAS_DELIMITER + DBConstants.NOTES),
+            rs.getBoolean(DBConstants.DDP_PARTICIPANT_RECORD_ALIAS + DBConstants.ALIAS_DELIMITER + DBConstants.MINIMAL_MR),
+            rs.getBoolean(
+                    DBConstants.DDP_PARTICIPANT_RECORD_ALIAS + DBConstants.ALIAS_DELIMITER + DBConstants.ABSTRACTION_READY),
+            rs.getString(DBConstants.DDP_PARTICIPANT_RECORD_ALIAS + DBConstants.ALIAS_DELIMITER
+                    + DBConstants.ADDITIONAL_VALUES_JSON),
+            (Long) rs.getObject(DBConstants.EXIT_DATE));
     }
 
-    public static Map<String, Participant> getParticipants(@NonNull String realm, String queryAddition) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/Participant.java:211]: `queryAddition` no longer used.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java [hunk 6]
```diff
@@ -48,12 +48,6 @@ public class DDPMedicalRecordDataRequest {
                     + "LEFT JOIN ddp_medical_record_log log on (log.medical_record_id = rec.medical_record_id) "
                     + "WHERE part.ddp_participant_id = ? AND part.ddp_instance_id = ? "
                     + "AND NOT rec.deleted <=> 1 AND rec.fax_sent is not null AND (log.type is null OR log.type = ?)";
-    private static final String SQL_SELECT_LOG_FOR_MEDICAL_RECORD =
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java:51]: Unused.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java [hunk 7]
```diff
@@ -117,18 +111,20 @@ public static List<Integer> writeInstitutionBundle(@NonNull Connection conn, @No
                     institutionRequest.getLastUpdated(), SystemUtil.SYSTEM)) {
                 medicalRecordIds = writeInstitutionInfo(conn, institutionRequest, instanceId, instanceName);
                 //participant lastVersion changed
-                Collection<Number> allMedicalRecordIds = getMedicalRecordIds(conn, ddpParticipantId, instanceId);
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java:120]: `Number` is incorrect. `Integer` is correct.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java [hunk 8]
```diff
@@ -218,27 +214,4 @@ public static Collection<Number> getMedicalRecordIds(Connection conn, String par
         }
         return medicalRecordIds;
     }
-
-    public boolean shouldHaveMedicalRecordLog(Connection conn, Number medicalRecordId) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/DDPMedicalRecordDataRequest.java:222]: Unused.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java [hunk 9]
```diff
@@ -83,44 +85,41 @@ public static int writeMedicalRecord(Connection conn, String institutionId) {
...
             }
             logger.info("Added new medical record for institution with id {}", institutionId);
+            return rs.getInt(1);
         } catch (SQLException e) {
             throw new DsmInternalError("Error inserting medical record for institution " + institutionId, e);
         }
-        return mrId;
     }
 
-    public static void writeNewRecordIntoDb(Connection conn, String query, String id, String instanceId) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java:108]: `id` is ambiguous.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java [hunk 10]
```diff
@@ -132,31 +131,32 @@ public static void writeNewRecordIntoDb(Connection conn, String query, String id
         }
     }
 
-    public static int writeInstitution(int participantId, @NonNull String ddpParticipantId, @NonNull String type,
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java:135]: `type` is ambiguous.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java [hunk 11]
```diff
@@ -132,31 +131,32 @@ public static void writeNewRecordIntoDb(Connection conn, String query, String id
...
             try (PreparedStatement insertInstitution = conn.prepareStatement(SQL_INSERT_INSTITUTION,
                     Statement.RETURN_GENERATED_KEYS)) {
                 insertInstitution.setString(1, ddpInstitutionId);
-                insertInstitution.setString(2, type);
+                insertInstitution.setString(2, institutionType);
                 insertInstitution.setInt(3, participantId);
                 insertInstitution.setLong(4, currentMilli);
                 insertInstitution.setLong(5, currentMilli);
                 int result = insertInstitution.executeUpdate();
-                // 1 (inserted) or 2 (updated) is good
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java:148]: This case (updated) can never actually happen.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java [hunk 12]
```diff
@@ -195,28 +196,29 @@ public static Integer writeInstitution(@NonNull Connection conn, @NonNull String
         return medicalRecordId;
     }
 
-    private static Integer createInstitutionMedicalRecord(@NonNull Connection conn, @NonNull PreparedStatement insertInstitution,
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java:198]: The result is never NULL, so primitive type can be used.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java [hunk 13]
```diff
@@ -195,28 +196,29 @@ public static Integer writeInstitution(@NonNull Connection conn, @NonNull String
...
-                                                          @NonNull String ddpParticipantId, String instanceName,
-                                                          String ddpInstitutionId, boolean updateElastic) {
+    private static int createInstitutionMedicalRecord(Connection conn, PreparedStatement insertInstitution,
+                                                      String ddpParticipantId, String instanceName,
+                                                      Institution institution, boolean updateElastic) {
+        String errorMsg = "Error getting new institution id for participant " + ddpParticipantId;
         try (ResultSet rs = insertInstitution.getGeneratedKeys()) {
-            if (rs.next()) { //no next if no generated return key -> update of institution timestamp does not return new key
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/MedicalRecordUtil.java:202]: This comment describes a case that should not happen, hence an exception.

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/elastic/migration/OsteoMigratorTest.java [hunk 14]
```diff
@@ -171,15 +172,15 @@ public void testOsteoMigrator() {
         verifyExportLogs(exportLogs);
 
         // should be onc history for OS1 only participant
-        String os1DdpPtpId = os1Ptp.getDdpParticipantIdOrThrow();
-        verifyOncHistory(os1DdpPtpId, List.of(os1Ptp.getParticipantIdOrThrow()));
+        String os1DdpPtpId = os1Ptp.getRequiredDdpParticipantId();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/elastic/migration/OsteoMigratorTest.java:175]: Opportunistically changing to the preferred method name `getRequiredDdpParticipantId`.

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/OncHistoryDetailPatchTest.java [hunk 15]
```diff
@@ -100,15 +110,23 @@ public void doPatchTest() {
             Profile esProfile = esParticipant.getProfile().orElseThrow();
             Assert.assertEquals(profile.getHruid(), esProfile.getHruid());
 
+            // the patch medical record is not written to ES
             Dsm dsm = esParticipant.getDsm().orElseThrow();
-            List<MedicalRecord> medRecords = dsm.getMedicalRecord();
-            log.debug("Med records {}", medRecords);
-            Assert.assertEquals(2, medRecords.size());
+            List<MedicalRecord> esMedRecords = dsm.getMedicalRecord();
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/OncHistoryDetailPatchTest.java:115]: Improved test to support refactoring in `OncHistoryDetailPatch`.

**Review by ssettipalli — APPROVED** [review]
looks good to me

**Pull request merged**